File: ocaml/lexer.ml
(* Lexer implementation.
 * Currently supports lexing basic arithmetic. *)

open Core

module M = Mark

(* Defines an exception if we encounter an invalid token. *)
exception InvalidToken of string
exception InvalidInt of string

(* Defines the legal operations. *)
type op = Pow | Plus | Minus | Times | Divide

(* Defines legal tokens. *)
type token =
  | Symbol of string
  | IntVal of int
  | Operator of op
  | Eq
  | LParen
  | RParen
  | Delim
  | Return
  | Eof

type mtoken = token M.t

(* Defines a custom lexer type. *)
type t = {
  fname : string;                   (* File name *)
  file : string;                    (* File contents *)
  length : int;                     (* Length of file contents *)
  pos : M.pos ref;                  (* Row, col position *)
  idx : int ref;                    (* Index into file contents *)
  saved_tok : mtoken option ref     (* Last token seen from peek *)
}

(* Stolen from my boy Henry. This takes a file name and returns the string
 * representation of the file. *)
let slurp fname = In_channel.with_file ~binary:false fname 
    ~f:(fun ch -> In_channel.input_all ch)

let file lexer = lexer.file
let fname lexer = lexer.fname

let string_of_op op =
  match op with
  | Pow -> "^"
  | Plus -> "+"
  | Minus -> "-"
  | Times -> "*"
  | Divide -> "/"

(* For debugging: converts a token into a user-readable string. *)
let string_of_token mtok =
  let tok = M.obj mtok in
  let tok_str = 
    match tok with
    | Symbol s -> "SYM " ^ s
    | IntVal i -> "" ^ string_of_int i
    | Operator op -> string_of_op op
    | Eq -> "="
    | LParen -> "("
    | RParen -> ")"
    | Delim -> ";"
    | Return -> "return"
    | Eof -> "EOF"
  in
  let (r1, c1) = M.start mtok in
  let (r2, c2) = M.stop mtok in
  tok_str ^ " (" ^ string_of_int r1 ^ ", " ^ string_of_int c1 ^ ")" ^ ", (" ^ string_of_int r2 ^ ", " ^ string_of_int c2 ^ ")"

(* Creates a new lexer. *)
let create fname = 
  let file = slurp fname in
  let result = {
    fname = fname;
    file = file;
    length = String.length file;
    pos = ref (1, 1);
    idx = ref 0;
    saved_tok = ref None;
  } in result

(* Fetches the next token from the lexer. *)
let rec next_token lxr = 
  let invalid_loc () =
    let r, c = !(lxr.pos) in
    Printf.sprintf "%s: line %d, character %d" lxr.fname r c
  in
  let incr_pos r c = lxr.idx := !(lxr.idx) + 1; lxr.pos := (r, c+1); in
  let incr_pos_new_row r = lxr.idx := !(lxr.idx) + 1; lxr.pos := (r+1, 1); in
  let get_ch () = String.get lxr.file !(lxr.idx) in
  let lxr_print_err = Err.print lxr.file lxr.fname in
  let result = 
    if !(lxr.idx) = lxr.length then M.create Eof !(lxr.pos) !(lxr.pos)
    else 
      let ch = get_ch () in
      let (r, c) = !(lxr.pos) in 
      let token = 
        match ch with
        (* Whitespace produces no token, so skip it. *)
        | '\n' -> 
          incr_pos_new_row r;
          next_token lxr
        | ' ' | '\t' | '\r' -> 
          incr_pos r c; 
          next_token lxr
        (* If you encounter a digit, gobble up all of the subsequent numbers *)
        | '0' .. '9' -> 
          let rec parse_digits num_str = 
            if !(lxr.idx) = lxr.length then num_str
            else
              let (r, c) = !(lxr.pos) in 
              let digit = get_ch () in
              let result = 
                match digit with
                | '0' .. '9' -> 
                  incr_pos r c;
                  parse_digits (num_str ^ Char.to_string digit)
                (* Ints should not have letters in them. *)
                | 'A' .. 'Z' | 'a' .. 'z' -> 
                  lxr_print_err (Mark.create digit !(lxr.pos) !(lxr.pos));
                  raise (InvalidInt (invalid_loc ()))
                | _ -> num_str
              in result
          in 
          let num = int_of_string (parse_digits "")
          in M.create (IntVal num) (r,c) !(lxr.pos)
        (* If you encounter a letter, interpret as a symbol or keyword. *)
        | 'A' .. 'Z' | 'a' .. 'z' ->
          let rec parse_symbol str =
            if !(lxr.idx) = lxr.length then str
            else
              let (r, c) = !(lxr.pos) in 
              let ch' = get_ch () in
              let result =
                match ch' with
                | 'A' .. 'Z' | 'a' .. 'z'
                | '0' .. '9' -> 
                  incr_pos r c;
                  parse_symbol (str ^ Char.to_string ch')
                | _ -> str
              in result
          in
          (* Handle return keyword *)
          let sym = parse_symbol "" in 
          let kwd = 
            (match sym with
             (* HERE: I suspect this is where we would extend to
               * support more keywords in the future. *)
             | "return" -> Return
             | _ -> Symbol sym)
          in M.create kwd (r,c) !(lxr.pos) 
        | _ -> 
          let t = 
            match ch with
            | '=' -> Eq
            | ';' -> Delim
            | '^' -> Operator Pow
            | '+' -> Operator Plus
            | '-' -> Operator Minus
            | '*' -> Operator Times
            | '/' -> Operator Divide
            | '(' -> LParen
            | ')' -> RParen
            | _ -> 
              lxr_print_err (M.create ch (r,c) !(lxr.pos));
              raise (InvalidToken (invalid_loc ()))
          in 
          incr_pos r c;
          M.create t (r,c) !(lxr.pos)
      in token
  in result

(* Pops the next token from the lexer. *)
let pop lxr =
  match !(lxr.saved_tok) with
  | Some tok -> lxr.saved_tok := None; tok
  | None -> next_token lxr

(* Peek at the next token in the lexer. *)
let peek lxr =
  match !(lxr.saved_tok) with
  | Some tok -> tok
  | None -> let result = next_token lxr in lxr.saved_tok := Some result; result

(* Drops the next token from the lexer. Equivalent to pop, but it ignores the
 * token. *)
let drop lxr = let (_ : mtoken) = pop lxr in ()

File: go/lexer/lexer.go
package lexer

import (
	"errors"
	"io/ioutil"
	"strconv"
	"unicode"
)

type OpType int
type TokenType int

const (
	// Op (operators)
	Pow OpType = iota
	Plus
	Minus
	Times
	Divide

	// Tok (tokens)
	Symbol TokenType = iota
	IntVal
	Operator
	Eq
	LParen
	RParen
	Delim
	Return
	Eof
)

type Token struct {
	Type TokenType
	Optr OpType
	Sym  string
	Val  int
}

type T struct {
	idx    int
	tokens []*Token
}

func (op OpType) ToString() string {
	switch op {
	case Pow:
		return "^"
	case Plus:
		return "+"
	case Minus:
		return "-"
	case Times:
		return "*"
	default:
		return "/"
	}
}

func (tok *Token) ToString() string {
	switch tok.Type {
	case Symbol:
		return tok.Sym
	case IntVal:
		return strconv.Itoa(tok.Val)
	case Operator:
		return tok.Optr.ToString()
	case Eq:
		return "="
	case LParen:
		return "("
	case RParen:
		return ")"
	case Delim:
		return ";"
	case Return:
		return "return"
	default:
		return "EOF"
	}
}

func newLexer(file string) (*T, error) {
	result := &T{}
	flength := len(file)

	var lexToTokens func(int) ([]*Token, error)
	lexToTokens = func(i int) ([]*Token, error) {
		if i == flength {
			return []*Token{&Token{Type: Eof}}, nil
		}

		ch := rune(file[i])

		var token *Token
		if unicode.IsSpace(ch) {
			return lexToTokens(i + 1)
		} else if unicode.IsDigit(ch) {
			temp := ""
			for i < flength {
				temp_ch := rune(file[i])
				if unicode.IsDigit(temp_ch) {
					temp += string(temp_ch)
				} else if unicode.IsLetter(temp_ch) {
					return nil, errors.New("invalid character encountered while lexing int")
				} else {
					break
				}
				i++
			}

			// HERE: i'm pretty sure this needs to be here so we lex the next character.
			i--

			tempInt, err := strconv.Atoi(temp)
			if err != nil {
				return nil, errors.New("failed to lex int")
			}
			token = &Token{Type: IntVal, Val: tempInt}
		} else if unicode.IsLetter(ch) {
			temp := ""
			for i < flength {
				temp_ch := rune(file[i])
				if unicode.IsDigit(temp_ch) || unicode.IsLetter(temp_ch) {
					temp += string(temp_ch)
				} else {
					break
				}
				i++
			}

			// HERE: i'm pretty sure this needs to be here so we lex the next character.
			i--

			// TODO: add more keyword checks here.
			switch temp {
			case "return":
				token = &Token{Type: Return}
			default:
				token = &Token{Type: Symbol, Sym: temp}
			}
		} else {
			switch ch {
			case '=':
				token = &Token{Type: Eq}
			case ';':
				token = &Token{Type: Delim}
			case '^':
				token = &Token{Type: Operator, Optr: Pow}
			case '+':
				token = &Token{Type: Operator, Optr: Plus}
			case '-':
				token = &Token{Type: Operator, Optr: Minus}
			case '*':
				token = &Token{Type: Operator, Optr: Times}
			case '/':
				token = &Token{Type: Operator, Optr: Divide}
			case '(':
				token = &Token{Type: LParen}
			case ')':
				token = &Token{Type: RParen}
			default:
				return nil, errors.New("unrecognized character in lexing")
			}
		}

		remainder, err := lexToTokens(i + 1)
		if err != nil {
			return nil, err
		}

		return append([]*Token{token}, remainder...), nil
	}

	result.idx = 0
	tokens, err := lexToTokens(0)
	if err != nil {
		return nil, err
	}

	result.tokens = tokens

	return result, nil
}

func NewLexer(filename string) (*T, error) {
	content, err := ioutil.ReadFile(filename)
	if err != nil {
		return nil, err
	}

	return newLexer(string(content))
}

func (lexer *T) Pop() *Token {
	if lexer.idx == len(lexer.tokens)-1 {
		return lexer.Peek()
	}
	result := lexer.Peek()
	lexer.idx++
	return result
}

func (lexer *T) Peek() *Token {
	return lexer.tokens[lexer.idx]
}

func (lexer *T) Drop() {
	lexer.Pop()
}
